pkg/volume/util/operationexecutor/operation_generator.go

```diff
@@ -1731,3 +1730,25 @@ func isDeviceOpened(deviceToDetach AttachedVolume, hostUtil hostutil.HostUtils)
 	}
 	return deviceOpened, nil
 }
+
+// findDetachablePluginBySpec is a variant of VolumePluginMgr.FindAttachablePluginByName() function.
+// The difference is that it bypass the CanAttach() check for CSI plugin, i.e. it assumes all CSI plugin supports detach.
+// The intention here is that a CSI plugin volume can end up in an Uncertain state,  so that a detach
+// operation will help it to detach no matter it actually has the ability to attach/detach.
+func findDetachablePluginBySpec(spec *volume.Spec, pm *volume.VolumePluginMgr) (volume.AttachableVolumePlugin, error) {
+	volumePlugin, err := pm.FindPluginBySpec(spec)
+	if err != nil {
+		return nil, err
+	}
+	if attachableVolumePlugin, ok := volumePlugin.(volume.AttachableVolumePlugin); ok {
+		if attachableVolumePlugin.GetPluginName() == "kubernetes.io/csi" {
```

Comment:
@saad-ali I thought it through and I think we don't actually need any other check(e.g. inline check) in the CanAttach() function because we are expecting to get a attachableVolumePlugin from the return. The CanAttach() should never return false to the caller.

That said, in this case, as long as it is a CSI plugin, it should be detachable no matter it is attachable or not. Please correct me if I am wrong.

Reply:
For CSI volume that are Ephemeral volume should not be attachable and therefore should not be detachable (which is checked by `CanAttach()`.

That said, if the proper checks happen on the attach side, detach may be less strict. So this should be fine though it makes me a little nervous.